### ISSUE TITLE
mode=aggressive not stopping Postfix brute-force attack

### ISSUE THREAD

**Issue by Spire9** [fail2ban/fail2ban#2876]
### Environment:
LAMP
Ubuntu 18.04
Fail2Ban v0.10.2
OpenSSH_7.6p1 Ubuntu-4ubuntu0.3, OpenSSL 1.0.2n 7 Dec 2017
- [x] Fail2Ban installed via OS/distribution mechanisms
- [ ] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

### The issue:
It appears someone is trying to brute-force into my email system.  I have mode=aggressive turned on for both postfix jails, but it sees nothing.  

### Configuration, dump and another helpful excerpts
```
[postfix]
enabled = true
findtime = 25h
bantime = 8d
maxretry = 1
mode = aggressive

[postfix-sasl]
enabled = true
findtime = 25h
bantime = 8d
maxretry = 1
mode = aggressive
```

#### Any customizations done to /etc/fail2ban/ configuration
None

#### Relevant parts of /var/log/fail2ban.log file:
Nothing

#### Relevant lines from monitored log files in question:
from /var/log/mail.log
```
Nov 20 23:30:11 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:30:15 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:30:33 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:30:38 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:30:44 www postfix/anvil[7312]: statistics: max connection rate 3/60s for (smtp:212.70.149.84) at Nov 20 23:21:34
Nov 20 23:30:44 www postfix/anvil[7312]: statistics: max connection count 1 for (smtp:212.70.149.84) at Nov 20 23:20:50
Nov 20 23:30:55 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:31:00 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:31:18 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:31:23 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:31:40 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:31:45 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:32:03 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:32:07 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:32:25 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:32:30 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:32:47 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:32:52 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:33:10 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:33:14 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:33:32 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:33:37 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:33:54 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:33:59 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:34:17 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
Nov 20 23:34:21 www postfix/smtpd[7309]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
```
same file, with Postfix set to -verbose
```
Nov 20 23:08:52 www postfix/smtpd[4419]: connection established
Nov 20 23:08:52 www postfix/smtpd[4419]: master_notify: status 0
Nov 20 23:08:52 www postfix/smtpd[4419]: name_mask: resource
Nov 20 23:08:52 www postfix/smtpd[4419]: name_mask: software
Nov 20 23:08:52 www postfix/smtpd[4419]: connect from unknown[212.70.149.84]
Nov 20 23:08:52 www postfix/smtpd[4419]: match_list_match: unknown: no match
Nov 20 23:08:52 www postfix/smtpd[4419]: match_list_match: 212.70.149.84: no match
Nov 20 23:08:52 www postfix/smtpd[4419]: match_list_match: unknown: no match
Nov 20 23:08:52 www postfix/smtpd[4419]: match_list_match: 212.70.149.84: no match
Nov 20 23:08:52 www postfix/smtpd[4419]: smtp_stream_setup: maxtime=300 enable_deadline=0
Nov 20 23:08:52 www postfix/smtpd[4419]: match_hostname: smtpd_client_event_limit_exceptions: unknown ~? 127.0.0.0/8
Nov 20 23:08:52 www postfix/smtpd[4419]: match_hostaddr: smtpd_client_event_limit_exceptions: 212.70.149.84 ~? 127.0.0.0/8
Nov 20 23:08:52 www postfix/smtpd[4419]: match_hostname: smtpd_client_event_limit_exceptions: unknown ~? [::ffff:127.0.0.0]/104
Nov 20 23:08:52 www postfix/smtpd[4419]: match_hostaddr: smtpd_client_event_limit_exceptions: 212.70.149.84 ~? [::ffff:127.0.0.0]/104
Nov 20 23:08:52 www postfix/smtpd[4419]: match_hostname: smtpd_client_event_limit_exceptions: unknown ~? [::1]/128
Nov 20 23:08:52 www postfix/smtpd[4419]: match_hostaddr: smtpd_client_event_limit_exceptions: 212.70.149.84 ~? [::1]/128
Nov 20 23:08:52 www postfix/smtpd[4419]: match_list_match: unknown: no match
Nov 20 23:08:52 www postfix/smtpd[4419]: match_list_match: 212.70.149.84: no match
Nov 20 23:08:52 www postfix/smtpd[4419]: send attr request = connect
Nov 20 23:08:52 www postfix/smtpd[4419]: send attr ident = smtp:212.70.149.84
Nov 20 23:08:52 www postfix/smtpd[4419]: private/anvil: wanted attribute: status
Nov 20 23:08:52 www postfix/smtpd[4419]: input attribute name: status
Nov 20 23:08:52 www postfix/smtpd[4419]: input attribute value: 0
Nov 20 23:08:52 www postfix/smtpd[4419]: private/anvil: wanted attribute: count
Nov 20 23:08:52 www postfix/smtpd[4419]: input attribute name: count
Nov 20 23:08:52 www postfix/smtpd[4419]: input attribute value: 1
Nov 20 23:08:52 www postfix/smtpd[4419]: private/anvil: wanted attribute: rate
Nov 20 23:08:52 www postfix/smtpd[4419]: input attribute name: rate
Nov 20 23:08:52 www postfix/smtpd[4419]: input attribute value: 3
Nov 20 23:08:52 www postfix/smtpd[4419]: private/anvil: wanted attribute: (list terminator)
Nov 20 23:08:52 www postfix/smtpd[4419]: input attribute name: (end)
Nov 20 23:08:52 www postfix/smtpd[4419]: report connect to all milters
Nov 20 23:08:52 www postfix/smtpd[4419]: milter_macro_lookup: "j"
Nov 20 23:08:52 www postfix/smtpd[4419]: milter_macro_lookup: result "www.example.com"
Nov 20 23:08:52 www postfix/smtpd[4419]: milter_macro_lookup: "{daemon_name}"
Nov 20 23:08:52 www postfix/smtpd[4419]: milter_macro_lookup: result "www.example.com"
Nov 20 23:08:52 www postfix/smtpd[4419]: milter_macro_lookup: "{daemon_addr}"
Nov 20 23:08:52 www postfix/smtpd[4419]: milter_macro_lookup: result "123.123.12.123"
Nov 20 23:08:52 www postfix/smtpd[4419]: milter_macro_lookup: "v"
Nov 20 23:08:52 www postfix/smtpd[4419]: milter_macro_lookup: result "Postfix 3.3.0"
Nov 20 23:08:52 www postfix/smtpd[4419]: milter8_connect: non-protocol events for protocol version 2: SMFIP_NOUNKNOWN SMFIP_NODATA
Nov 20 23:08:52 www postfix/smtpd[4419]: milter8_connect: transport=inet endpoint=localhost:12301
Nov 20 23:08:52 www postfix/smtpd[4419]: trying... [127.0.0.1]
Nov 20 23:08:52 www postfix/smtpd[4419]: vstream_tweak_tcp: TCP_MAXSEG 32741
Nov 20 23:08:52 www postfix/smtpd[4419]: fd=12: stream buffer size old=0 new=65482
Nov 20 23:08:52 www postfix/smtpd[4419]: milter8_connect: my_version=0x2
Nov 20 23:08:52 www postfix/smtpd[4419]: milter8_connect: my_actions=0x1ff SMFIF_ADDHDRS SMFIF_CHGBODY SMFIF_ADDRCPT SMFIF_DELRCPT SMFIF_CHGHDRS SMFIF_QUARANTINE SMFIF_CHGFROM SMFIF_ADDRCPT_PAR SMFIF_SETSYMLIST
Nov 20 23:08:52 www postfix/smtpd[4419]: milter8_connect: my_events=0x7f SMFIP_NOCONNECT SMFIP_NOHELO SMFIP_NOMAIL SMFIP_NORCPT SMFIP_NOBODY SMFIP_NOHDRS SMFIP_NOEOH
Nov 20 23:08:52 www postfix/smtpd[4419]: milter8_connect: milter inet:localhost:12301 version 2
Nov 20 23:08:52 www postfix/smtpd[4419]: milter8_connect: events SMFIP_NOHELO
Nov 20 23:08:52 www postfix/smtpd[4419]: milter8_connect: requests SMFIF_ADDHDRS SMFIF_CHGHDRS SMFIF_SETSYMLIST
Nov 20 23:08:52 www postfix/smtpd[4419]: milter8_conn_event: milter inet:localhost:12301: connect unknown/212.70.149.84
Nov 20 23:08:52 www postfix/smtpd[4419]: event: SMFIC_CONNECT; macros: j=www.example.com {daemon_name}=www.example.com {daemon_addr}=123.123.12.123 v=Postfix 3.3.0
Nov 20 23:08:52 www postfix/smtpd[4419]: reply: SMFIR_CONTINUE data 0 bytes
Nov 20 23:08:52 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 220 www.example.com ESMTP Postfix (Ubuntu)
Nov 20 23:08:53 www postfix/smtpd[4419]: < unknown[212.70.149.84]: EHLO User
Nov 20 23:08:53 www postfix/smtpd[4419]: report helo to all milters
Nov 20 23:08:53 www postfix/smtpd[4419]: milter_macro_lookup: "{tls_version}"
Nov 20 23:08:53 www postfix/smtpd[4419]: milter_macro_lookup: "{cipher}"
Nov 20 23:08:53 www postfix/smtpd[4419]: milter_macro_lookup: "{cipher_bits}"
Nov 20 23:08:53 www postfix/smtpd[4419]: milter_macro_lookup: "{cert_subject}"
Nov 20 23:08:53 www postfix/smtpd[4419]: milter_macro_lookup: "{cert_issuer}"
Nov 20 23:08:53 www postfix/smtpd[4419]: milter8_helo_event: milter inet:localhost:12301: helo User
Nov 20 23:08:53 www postfix/smtpd[4419]: event: SMFIC_HELO; macros: (none)
Nov 20 23:08:53 www postfix/smtpd[4419]: skipping event SMFIC_HELO for milter inet:localhost:12301
Nov 20 23:08:53 www postfix/smtpd[4419]: match_list_match: unknown: no match
Nov 20 23:08:53 www postfix/smtpd[4419]: match_list_match: 212.70.149.84: no match
Nov 20 23:08:53 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250-www.example.com
Nov 20 23:08:53 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250-PIPELINING
Nov 20 23:08:53 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250-SIZE 52428800
Nov 20 23:08:53 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250-VRFY
Nov 20 23:08:53 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250-ETRN
Nov 20 23:08:53 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250-STARTTLS
Nov 20 23:08:53 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250-ENHANCEDSTATUSCODES
Nov 20 23:08:53 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250-8BITMIME
Nov 20 23:08:53 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250-DSN
Nov 20 23:08:53 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250 SMTPUTF8
Nov 20 23:08:54 www postfix/smtpd[4419]: < unknown[212.70.149.84]: RSET
Nov 20 23:08:54 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 250 2.0.0 Ok
Nov 20 23:08:55 www postfix/smtpd[4419]: < unknown[212.70.149.84]: AUTH LOGIN
Nov 20 23:08:55 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 503 5.5.1 Error: authentication not enabled
Nov 20 23:08:57 www postfix/smtpd[4419]: < unknown[212.70.149.84]: QUIT
Nov 20 23:08:57 www postfix/smtpd[4419]: > unknown[212.70.149.84]: 221 2.0.0 Bye
Nov 20 23:08:57 www postfix/smtpd[4419]: match_hostname: smtpd_client_event_limit_exceptions: unknown ~? 127.0.0.0/8
Nov 20 23:08:57 www postfix/smtpd[4419]: match_hostaddr: smtpd_client_event_limit_exceptions: 212.70.149.84 ~? 127.0.0.0/8
Nov 20 23:08:57 www postfix/smtpd[4419]: match_hostname: smtpd_client_event_limit_exceptions: unknown ~? [::ffff:127.0.0.0]/104
Nov 20 23:08:57 www postfix/smtpd[4419]: match_hostaddr: smtpd_client_event_limit_exceptions: 212.70.149.84 ~? [::ffff:127.0.0.0]/104
Nov 20 23:08:57 www postfix/smtpd[4419]: match_hostname: smtpd_client_event_limit_exceptions: unknown ~? [::1]/128
Nov 20 23:08:57 www postfix/smtpd[4419]: match_hostaddr: smtpd_client_event_limit_exceptions: 212.70.149.84 ~? [::1]/128
Nov 20 23:08:57 www postfix/smtpd[4419]: match_list_match: unknown: no match
Nov 20 23:08:57 www postfix/smtpd[4419]: match_list_match: 212.70.149.84: no match
Nov 20 23:08:57 www postfix/smtpd[4419]: send attr request = disconnect
Nov 20 23:08:57 www postfix/smtpd[4419]: send attr ident = smtp:212.70.149.84
Nov 20 23:08:57 www postfix/smtpd[4419]: private/anvil: wanted attribute: status
Nov 20 23:08:57 www postfix/smtpd[4419]: input attribute name: status
Nov 20 23:08:57 www postfix/smtpd[4419]: input attribute value: 0
Nov 20 23:08:57 www postfix/smtpd[4419]: private/anvil: wanted attribute: (list terminator)
Nov 20 23:08:57 www postfix/smtpd[4419]: input attribute name: (end)
Nov 20 23:08:57 www postfix/smtpd[4419]: abort all milters
Nov 20 23:08:57 www postfix/smtpd[4419]: milter8_abort: abort milter inet:localhost:12301
Nov 20 23:08:57 www postfix/smtpd[4419]: disconnect event to all milters
Nov 20 23:08:57 www postfix/smtpd[4419]: milter8_disc_event: quit milter inet:localhost:12301
Nov 20 23:08:57 www postfix/smtpd[4419]: disconnect from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4
Nov 20 23:08:57 www postfix/smtpd[4419]: free all milters
Nov 20 23:08:57 www postfix/smtpd[4419]: free milter inet:localhost:12301
Nov 20 23:08:57 www postfix/smtpd[4419]: master_notify: status 1
Nov 20 23:08:57 www postfix/smtpd[4419]: connection closed
```

**Comment by Spire9** [comment]
I'm not sure if this is a bug or a new feature request...  I can't read the advanced regex used in fail2ban jails well enough to debug this myself.  
The attacker was guessing at a rate of about 4320 per day on my mail server, all from the same IP.  
The behavior that is needed would be for the following pattern to get picked up as a authentication failure:
```
^.*  auth=0/ .*$
```

**Comment by sebres** [comment]
Well, it works pretty well in newer version:
<details><summary>test fail2ban-regex with postfix[mode=aggressive] ...</summary>

```
$ fail2ban-regex -vv /tmp/f2b.txt postfix[mode=aggressive]

Running tests
=============

Use   failregex filter file : postfix, basedir: /etc/fail2ban/
Use   filter options : {'mode': 'aggressive'}
Real  filter options : {'logtype': 'file', 'known/journalmatch': '_SYSTEMD_UNIT=postfix.service', 'journalmatch': '_SYSTEMD_UNIT=postfix.service', 'mode': 'aggressive', 'datepattern': '{^LN-BEG}'}
Use      datepattern : {^LN-BEG} : Default Detectors
Use         log file : /tmp/f2b.txt
Use         encoding : UTF-8


Results
=======

Prefregex: 12 total
|  ^(?:\[\])?\s*(?:<[^.]+\.[^.]+>\s+)?(?:\S+\s+)?(?:kernel:\s?\[ *\d+\.\d+\]:?\s+)?(?:@vserver_\S+\s+)?(?:(?:(?:\[\d+\])?:\s+[\[\(]?postfix(-\w+)?/\w+(?:/smtp[ds])?(?:\(\S+\))?[\]\)]?:?|[\[\(]?postfix(-\w+)?/\w+(?:/smtp[ds])?(?:\(\S+\))?[\]\)]?:?(?:\[\d+\])?:?)\s+)?(?:\[ID \d+ \S+\]\s+)?(?:warning:|(?:\w+: reject:|(?:improper command pipelining|too many errors) after \S+)|(?:lost connection after(?! DATA) [A-Z]+|disconnect(?= from \S+(?: \S+=\d+)* auth=0/(?:[1-9]|\d\d+)))) (?P<content>.+)$
|      {'content': u'from unknown[212.70.149.84] ehlo=1 auth=0/1 rset=1 quit=1 commands=3/4'}
|      ...
`-

Failregex: 12 total
|-  #) [# of hits] regular expression
|   ...
|   8) [12] ^from [^[]*\[<HOST>\](?::\d+)?:?
|      212.70.149.84  Fri Nov 20 23:30:15 2020
|      212.70.149.84  Fri Nov 20 23:30:38 2020
|      212.70.149.84  Fri Nov 20 23:31:00 2020
|      212.70.149.84  Fri Nov 20 23:31:23 2020
|      212.70.149.84  Fri Nov 20 23:31:45 2020
|      212.70.149.84  Fri Nov 20 23:32:07 2020
|      212.70.149.84  Fri Nov 20 23:32:30 2020
|      212.70.149.84  Fri Nov 20 23:32:52 2020
|      212.70.149.84  Fri Nov 20 23:33:14 2020
|      212.70.149.84  Fri Nov 20 23:33:37 2020
|      212.70.149.84  Fri Nov 20 23:33:59 2020
|      212.70.149.84  Fri Nov 20 23:34:21 2020
`-

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [26] {^LN-BEG}(?:DAY )?MON Day %k:Minute:Second(?:\.Microseconds)?(?: ExYear)?
|  ...
`-

Lines: 26 lines, 0 ignored, 12 matched, 14 missed
[processed in 0.00 sec]

|- Missed line(s):
|  Nov 20 23:30:11 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:30:33 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:30:44 www postfix/anvil[7312]: statistics: max connection rate 3/60s for (smtp:212.70.149.84) at Nov 20 23:21:34
|  Nov 20 23:30:44 www postfix/anvil[7312]: statistics: max connection count 1 for (smtp:212.70.149.84) at Nov 20 23:20:50
|  Nov 20 23:30:55 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:31:18 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:31:40 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:32:03 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:32:25 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:32:47 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:33:10 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:33:32 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:33:54 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
|  Nov 20 23:34:17 www postfix/smtpd[7309]: connect from unknown[212.70.149.84]
`-
```
</details>

I guess it was fixed in #2630. This will be released with 0.10.6 (0.11.2) soon.
Since known delayed delivery or non-delivery of updates on debians, I'll also try to provide a deb-package together with the release.

**Comment by Spire9** [comment]
Thanks sebres.  It's much appreciated.
